depinject/internal/codegen/type_test.go

```diff
@@ -0,0 +1,80 @@
+package codegen
+
+import (
+	"bytes"
+	"go/ast"
+	"go/printer"
+	"go/token"
+	"reflect"
+	"testing"
+
+	"gotest.tools/v3/assert"
+
+	"cosmossdk.io/depinject/internal/graphviz"
+)
+
+type MyInt int
+
+type AStruct struct {
+	Foo int
+}
+
+type AStructWrapper AStruct
+
+type AnInterface interface{}
+
+func TestTypeExpr(t *testing.T) {
+	expectTypeExpr(t, false, "bool")
+	expectTypeExpr(t, uint(0), "uint")
+	expectTypeExpr(t, uint8(0), "uint8")
+	expectTypeExpr(t, uint16(0), "uint16")
+	expectTypeExpr(t, uint32(0), "uint32")
+	expectTypeExpr(t, uint64(0), "uint64")
+	expectTypeExpr(t, int(0), "int")
+	expectTypeExpr(t, int8(0), "int8")
+	expectTypeExpr(t, int16(0), "int16")
+	expectTypeExpr(t, int32(0), "int32")
+	expectTypeExpr(t, int64(0), "int64")
+	expectTypeExpr(t, float32(0), "float32")
+	expectTypeExpr(t, float64(0), "float64")
+	expectTypeExpr(t, complex64(0), "complex64")
+	expectTypeExpr(t, complex128(0), "complex128")
+	expectTypeExpr(t, MyInt(0), "codegen.MyInt")
+	expectTypeExpr(t, [1]int{0}, "[1]int")
+	expectTypeExpr(t, []int{}, "[]int")
+	expectTypeExpr(t, make(chan int), "chan int")
+	expectTypeExpr(t, make(<-chan int), "<-chan int")
+	expectTypeExpr(t, make(chan<- int), "chan<- int")
+	expectTypeExpr(t, func(int, string) (bool, error) { return false, nil },
+		"func(int, string) (bool, error)",
+	)
+	expectTypeExpr(t, func(int, ...string) (bool, error) { return false, nil },
+		"func(int, ...string) (bool, error)",
+	)
+	expectTypeExpr(t, AStruct{}, "codegen.AStruct")
+	expectTypeExpr(t, map[string]graphviz.Attributes{}, "map[string]graphviz.Attributes")
+	expectTypeExpr(t, &AStruct{}, "*codegen.AStruct")
+	expectTypeExpr(t, AStructWrapper{}, "codegen.AStructWrapper")
+	expectTypeExpr(t, "abc", "string")
+	expectTypeExpr(t, uintptr(0), "uintptr")
+	expectTypeExpr(t, (*AnInterface)(nil), "*codegen.AnInterface")
```

Comment:
```suggestion
	expectTypeExpr(t, (*AnInterface)(nil), "*codegen.AnInterface")
	expectTypeExpr(t, &BStruct[int]{}, "*codegen.BStruct[int]")
```

If we add all the types here, we could add a struct with generics like `BStruct[T any] struct`

Reply:
Supporting generic types was a bit non-trivial because they're not supported in the `reflect.Type` API explicitly but I need to manually parse the type params. Can you take another look @julienrbrt @kocubinski and let me know if this approach looks okay?